vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#pragma once
+#include "vw/core/array_parameters.h"
+#include "vw/core/example.h"
+#include "vw/core/vw_fwd.h"
+
+#include <ctime>
+#include <list>
+#include <unordered_map>
+#include <vector>
+
+namespace VW
+{
+namespace reductions
+{
+VW::LEARNER::base_learner* eigen_memory_tree_setup(VW::setup_base_i& stack_builder);
+
+namespace eigen_memory_tree
+{
+
+enum class emt_scorer_type
+{
+  random = 1,
+  distance = 2,
+  self_consistent_rank = 3,
+  not_self_consistent_rank = 4
+};
+
+enum class emt_router_type
+{
+  random = 1,
+  eigen = 2
+};
+
+enum class emt_scorer_feature_type
+{
+  abs_diff,
+  diff
+};
+
+enum class emt_scorer_example_type
+{
+  self_consistent,
+  not_self_consistent
+};
+
+
+struct emt_example
+{
+  VW::flat_example* base;  // base example only includes the base features without interaction flags
+  VW::flat_example* full;  // full example includes the interactions that were passed in as flags
+  uint32_t label;
+  uint32_t tag;
+  float score;
+
+  emt_example();
+  emt_example(VW::workspace& all, VW::example* ex);
+};
+
+struct emt_LRU
+{
+  using K = emt_example*;
+  using V = std::list<K>::iterator;
+
+  std::list<K> list;
+  std::unordered_map<K, V> map;
+
+  int max_size;
+
+  emt_LRU(unsigned long max_size);
+  K bound(K item);
+};
+
+struct emt_node
+{
+  emt_node* parent; // parent node
```

Comment:
are we using parent anywhere?

Reply:
Good catch. No I don't think we are. I think I just added that out of habit.

Reply:
I've cleaned this up.